### PR TITLE
Allow DataGenerator for ActivationDefence

### LINKED ISSUE #373: Poisoning defense with not in-memory dataset
**Is your feature request related to a problem? Please describe.**
Most (if not all) poisoning filtering defenses, such as `ActivationDefense`, require the entire dataset to be passed in, in-memory. This doesn't work for larger datasets or on memory-constrained systems for moderate sized datasets.

**Describe the solution you'd like**
The ability to pass in a `DataGenerator` instead of a full dataset.

### PR DESCRIPTION
**Pull request by ebubae**
# Description

This PR allows the activation clustering defense to accept the ART `DataGenerator` type instead of passing a full `x_train, y_train` numpy arrays on initialization. This allows the defense to be used in cases where the entire dataset doesn't fit in memory.

Fixes #373 

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Tests were added to `test_activation_defense` for generators

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/defences/detector/poison/test_activation_defence.py [hunk 1]
```diff
@@ -17,11 +17,14 @@
 # SOFTWARE.
 from __future__ import absolute_import, division, print_function, unicode_literals
 
+import keras
 import logging
 import unittest
 
 import numpy as np
+from keras_preprocessing.image import ImageDataGenerator
```
Comment (beat-buesser) [inline tests/defences/detector/poison/test_activation_defence.py:25]:
```suggestion
import keras
from keras_preprocessing.image import ImageDataGenerator
```

File: tests/defences/detector/poison/test_activation_defence.py [hunk 2]
```diff
@@ -17,11 +17,14 @@
 # SOFTWARE.
 from __future__ import absolute_import, division, print_function, unicode_literals
 
+import keras
```
Comment (beat-buesser) [inline tests/defences/detector/poison/test_activation_defence.py:20]:
```suggestion
```

File: tests/defences/detector/poison/test_activation_defence.py [hunk 3]
```diff
@@ -114,82 +126,122 @@ def test_detect_poison(self):
...
         self.assertNotEqual(sum_dist, sum_size)
+        self.assertNotEqual(sum_dist_gen, sum_size_gen)
 
     def test_evaluate_defense(self):
         # Get MNIST
         (x_train, _), (_, _), (_, _) = self.mnist
 
         kwargs = {"nb_clusters": 2, "nb_dims": 10, "reduce": "PCA"}
         _, _ = self.defence.detect_poison(**kwargs)
+        # _, _ = self.defence_gen.detect_poison(**kwargs)
```
Comment (beat-buesser) [inline tests/defences/detector/poison/test_activation_defence.py:182]: Is this line still required?
```suggestion
        # _, _ = self.defence_gen.detect_poison(**kwargs)
```

File: tests/defences/detector/poison/test_activation_defence.py (hunk 1 above)
Comment (ebubae) [reply tests/defences/detector/poison/test_activation_defence.py:23]: addressed in commit

File: tests/defences/detector/poison/test_activation_defence.py (hunk 2 above)
Comment (ebubae) [reply tests/defences/detector/poison/test_activation_defence.py]: addressed in commit

File: tests/defences/detector/poison/test_activation_defence.py (hunk 3 above)
Comment (ebubae) [reply tests/defences/detector/poison/test_activation_defence.py]: un-commented in commit

**Comment by beat-buesser** [comment]
@ebubae Thank you very much for adding support for data generators on the activation defense!

**Pull request merged**